openedx/openedx-platform
# [BD-14] fix: call blockstore APIs in atomic transactions.

## Description

Ensures data integrity of content library views when accessing blockstore as an app, running on a separate database from edxapp.

### Background

[edx-platform#29779](https://github.com/openedx/edx-platform/pull/29779) moved the blockstore app into edx-platform. On edx.org, blockstore was being run as a separate IDA, and therefore used its own database. Now that they're being run together under the same service umbrella, the blockstore database is configured as a secondary database in Django settings, and there's a [database router](https://github.com/openedx/edx-platform/blob/master/openedx/core/lib/blockstore_api/db_routers.py) which tells the service which database to use for which model requests.

When blockstore runs on the IDA, it has `ATOMIC_REQUESTS: true` enabled so that [all views are encased in a single transaction](https://docs.djangoproject.com/en/4.0/topics/db/transactions/). This works great when you only have a single database running for a service, and you want all the views for that service to be atomic. But when you have multiple databases, if [_any_ database has ATOMIC_REQUESTS:true, then _all views_ are made atomic](https://github.com/django/django/blob/4f92cf87b013801810226928ddd20097f6e4fccf/django/core/handlers/base.py#L345-L354), not just the ones that use this database. This essentially adds 2 unnecessary database queries per request. The edxapp database can probably handle this – it fields 100x more requests per second than the blockstore database does. But the blockstore database is small, and [runs on shared architecture](https://openedx.slack.com/archives/C02QN50TYMD/p1651247015947389?thread_ts=1650530641.419139&cid=C02QN50TYMD), and so if each edxapp view creates a transaction on the blockstore database, it'll fall over.

But we need the blockstore views to be atomic to ensure data integrity. So, before we enable blockstore-as-an-app in edx.org production, we need to wrap each of the Content Library REST API views which use the Blockstore API views with the atomic decorator as shown here.

This change is a data integrity change only, and should not visibly affect the user experience.

## Supporting information

See:

* https://github.com/openedx/blockstore/pull/149#issuecomment-1079223585
* [Slack comments starting here](https://openedx.slack.com/archives/C02QN50TYMD/p1651244167374359?thread_ts=1650530641.419139&cid=C02QN50TYMD)

## Testing instructions

Sandbox:

* LMS: https://pr30456.sandbox.opencraft.hosting/
* Studio: https://studio.pr30456.sandbox.opencraft.hosting/

This change is covered by the updated tests, but it's a good idea to ensure that Content Libraries still work as expected.

1. [Login to the LMS](https://pr30456.sandbox.opencraft.hosting/login) as a staff member (either `staff@example.com` / `edx` or `edx@example.com` / `edx` will work)
2. [Login to Studio](https://studio.pr30456.sandbox.opencraft.hosting/signin), and visit the Libraries tab.
3. Add a new content library.
4. Add an block, and upload a file.
5. Save/Publish the block, and ensure this works as expected.

To see the collections/bundles/drafts created, you can:
1. [Login to the Django Admin](https://pr30456.sandbox.opencraft.hosting/admin/) as a superuser (`edx@example.com` / `edx`)
2. Visit the [Blockstore Bundles](https://pr30456.sandbox.opencraft.hosting/admin/bundles/) section and browse around.

## Deadline

ASAP -- we need this change in place before we start using the blockstore app in edx.org production.

## Other information

This change should also be backported to Nutmeg.

FYI @ormsbee @symbolist @bszabo
